### ISSUE TITLE
NTI gives incorrect error "property never defined on this"

### ISSUE THREAD

**Issue by myphysicslab** [google/closure-compiler#2426]
A common pattern I use in many classes fails on one particular class when using NTI.  This is from the constructor for the class:

```javascript
  this.addParameter(new ParameterNumber(this, TabLayout.en.SIM_WIDTH,
      TabLayout.i18n.SIM_WIDTH, this.getSimWidth, this.setSimWidth));
```
The errors are
```
src/sims/common/TabLayout.js:365: WARNING - Property getSimWidth never defined on this
      TabLayout.i18n.SIM_WIDTH, this.getSimWidth, this.setSimWidth));
                                ^^^^^^^^^^^^^^^^

src/sims/common/TabLayout.js:365: WARNING - Property setSimWidth never defined on this
      TabLayout.i18n.SIM_WIDTH, this.getSimWidth, this.setSimWidth));
                                                  ^^^^^^^^^^^^^^^^
```
Those properties (methods) are in fact defined later in the file, for example:

```javascript
/** Returns the width of the simulation LabCanvas, as fraction of available width
@return {number} width of the simulation LabCanvas, as fraction of available width
*/
TabLayout.prototype.getSimWidth = function() {
  return this.simWidth_;
};
```

I discovered this back in September 2016 and found a workaround at that time:

```javascript
  // using 'self' fixes several NTI compiler errors
  var self = /** @type {myphysicslab.sims.common.TabLayout}*/(this);
  this.addParameter(new ParameterNumber(this, TabLayout.en.SIM_WIDTH,
      TabLayout.i18n.SIM_WIDTH, self.getSimWidth, self.setSimWidth));
```

But when using `BanNullDeref` conformance rule that results in warnings:

```
src/sims/common/TabLayout.js:365: WARNING - Violation: Possible dereference of null value.
      TabLayout.i18n.SIM_WIDTH, self.getSimWidth, self.setSimWidth));
                                ^^^^^^^^^^^^^^^^
```

I can't see why only this class, out of probably 100, is having this problem.  It seems to compile OK even with the warnings.  And OTI has no warnings.  The constructor has a bit more going on in it than others, for example defining more functions.  I tried removing the most complicated function definition from the constructor to simplify it.

You can see the [source code here](https://github.com/myphysicslab/myphysicslab/blob/master/src/sims/common/TabLayout.js#L364).

I could try cloning this class and stripping away functionality until the "property never defined on this" error goes away.

**Comment by dimvar** [comment]
You can still use the workaround, just make sure that the type in the cast is non-nullable (add an exclamation mark).

**Comment by dimvar** [comment]
I tried a simple example and I get the expected warning.

> I could try cloning this class and stripping away functionality until the "property never defined on this" error goes away.

That would be helpful. Without a repro, I can't do much.

**Comment by myphysicslab** [comment]
I've adapted my code to deal with #2430.  Still have this problem of "property never defined on this" for this only this one class.

Using the "assign to null" trick to see what the compiler is interesting.  Here is what I see in a similar place in another class's constructor, it is the correct type:
```
src/lab/view/SimView.js:229: WARNING - The right side in the assignment is not a subtype
of the left side.
Expected : null
Found    : myphysicslab.lab.view.SimView{memorizables_:Array<?>, scaleTogether_:boolean}

  var /** null */n = this;
                     ^^^^
```
But for TabLayout we get:
```
src/sims/common/TabLayout.js:362: WARNING - The right side in the assignment is not a subtype of
the left side.
Expected : null
Found    : {controls_:Array<?>, debug_layout:boolean, div_term:{style:{border:string} (loose)} (loose),
 div_time_graph:Element, graph_controls:{style:{border:string} (loose)} (loose), limitSize_:boolean, 
redoLayout:LOOSE_TOP_FUNCTION, show_term_cb:?, sim_controls:{style:{border:string, 
marginLeft:string} (loose)} (loose), terminalEnabled_:boolean, 
timeGraphCanvas:myphysicslab.lab.view.LabCanvas, time_graph_controls:{style:{border:string} 
(loose)} (loose)} (loose)

  var /** null */n = this;
                     ^^^^
```
I'm looking for typos but not finding anything.  The application gets built and this class seems to be working OK (which argues against a typo).

**Comment by dimvar** [comment]
So, the problem is that somehow the type of `this` becomes a loose object, instead of `TabLayout`. I'm not sure why.

Another thing you can try for more debugging insights, is to do the assignment `var /** null */n = this;` in various places inside the constructor. That will tell us whether the type of `this` was wrong from the start, or if it became wrong later in the constructor, after some bad type operation.

Also, I hadn't realized your project is open source. If you can tell me how to compile it with NTI, then I can try to repro there, without a small example.

**Comment by myphysicslab** [comment]
I've found the code that is causing the problem.  The constructor is 200 lines long, so if you want to see the entire constructor see the [source code](https://github.com/myphysicslab/myphysicslab/blob/master/src/sims/common/TabLayout.js#L167). 

This boolean constant is defined early on:

```javascript
  /** whether to put dashed borders around elements for debugging layout
  * @type {boolean}
  * @const
  */
  this.debug_layout = false;
```

After any "if" statement using `debug_layout` the type of `this` becomes a loose object.  For example:

```javascript
  /** @type {!Element} */
  this.div_contain = TabLayout.getElementById(elem_ids, 'container');
  if (this.debug_layout) {
    this.div_contain.style.border = 'dashed 1px red';
  }
```
By removing all those "if" statements involving `debug_layout` the error goes away.

If it's still of interest I can get you instructions for building my project with NTI.  I'm updating some things for the latest compiler and library changes but will be done soon.

**Comment by dimvar** [comment]
Got a repro! Thanks for the help, will look into fixing this.

````
/** @constructor */
function Foo(x) {
  var /** null */ n;

  /** @type {boolean} */
  this.mybool = false;

  if (this.mybool) {
    this.prop1 = 123;
  }

  n = this; // loose here
}
````

**Comment by ChadKillingsworth** [comment]
This really explains a lot for some of the things I've seen the last two days. It's hitting me hard on generics. I hadn't figured out a root cause yet though.

**Comment by dimvar** [comment]
OK, after I submit a fix for this, let me know if you still see the weird warning, and then you can file a separate issue for that.

**Comment by dimvar** [comment]
Fixed by a003e52aecf8347be3038a992bf6340569c5afc5
